x-pack/plugins/security/server/authentication/providers/pki.ts

```diff
@@ -255,23 +261,53 @@ export class PKIAuthenticationProvider extends BaseAuthenticationProvider {
    */
   private getCertificateChain(peerCertificate: DetailedPeerCertificate | null) {
     const certificateChain = [];
+    const certificateStrings = [];
+    let isChainIncomplete = false;
     let certificate: DetailedPeerCertificate | null = peerCertificate;
-    while (certificate !== null && Object.keys(certificate).length > 0) {
+
+    while (certificate && Object.keys(certificate).length > 0) {
       certificateChain.push(certificate.raw.toString('base64'));
+      certificateStrings.push(this.stringifyCertificate(certificate));
 
       // For self-signed certificates, `issuerCertificate` may be a circular reference.
       if (certificate === certificate.issuerCertificate) {
         this.logger.debug('Self-signed certificate is detected in certificate chain');
-        certificate = null;
+        break;
+      } else if (certificate.issuerCertificate === undefined) {
+        // The chain is only considered to be incomplete if one or more issuerCertificate values is undefined;
+        // this is not an expected return value from Node, but it can happen in some edge cases
+        isChainIncomplete = true;
+        break;
       } else {
+        // Repeat the loop
         certificate = certificate.issuerCertificate;
       }
     }
 
-    this.logger.debug(
-      `Peer certificate chain consists of ${certificateChain.length} certificates.`
-    );
+    this.logger.debug(`Peer certificate chain: [${certificateStrings.join(', ')}]`);
+
+    return { certificateChain, isChainIncomplete };
+  }
+
+  private stringifyCertificate(peerCertificate: DetailedPeerCertificate) {
+    const {
+      subject,
+      issuer,
+      issuerCertificate,
+      subjectaltname,
+      valid_from: validFrom,
+      valid_to: validTo,
+    } = peerCertificate;
+
+    let issuerCertType: string;
```

Comment:
__nit:__ would you mind adding a comment here explaining why we do this and what every "type" means?

Reply:
Done in f2c3f3c.

FWIW, I don't *think* there is ever a case where a peer certificate has a null `issuerCertificate` value. At least, I haven't been able to reproduce that locally.

_However_, the way the `getCertificateChain` method was written previously, if the authentication process made it to this step and the peer certificate was valid but had a null `issuerCertificate` value, Kibana would allow the authentication attempt to proceed. I didn't want to change that behavior and potentially introduce a regression, but I can't be sure if we would encounter a null value or not, seeing as we are encountering undefined values when we shouldn't be.